application/src/main/java/org/togetherjava/tjbot/commands/moderation/scam/ScamBlocker.java

```diff
@@ -0,0 +1,363 @@
+
+        reportChannelPattern = config.getScamBlocker().getReportChannelPattern();
+        Predicate<String> isReportChannelName =
+                Pattern.compile(reportChannelPattern).asMatchPredicate();
+        isReportChannel = channel -> isReportChannelName.test(channel.getName());
+        hasRequiredRole = Pattern.compile(config.getSoftModerationRolePattern()).asMatchPredicate();
+    }
+
+    @Override
+    public @NotNull String getName() {
+        return "scam-blocker";
+    }
+
+    @Override
+    public void onSelectionMenu(@NotNull SelectionMenuEvent event, @NotNull List<String> args) {
+        throw new UnsupportedOperationException("Not used");
+    }
+
+    @Override
+    public void acceptComponentIdGenerator(@NotNull ComponentIdGenerator generator) {
+        componentIdGenerator = generator;
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull GuildMessageReceivedEvent event) {
+        if (event.getAuthor().isBot() || event.isWebhookMessage()) {
+            return;
+        }
+
+        if (mode == ScamBlockerConfig.Mode.OFF) {
+            return;
+        }
+
+        Message message = event.getMessage();
+        String content = message.getContentDisplay();
+        if (!scamDetector.isScam(content)) {
+            return;
+        }
+
+        if (scamHistoryStore.hasRecentScamDuplicate(message)) {
+            takeActionWasAlreadyReported(event);
+            return;
+        }
+
+        takeAction(event);
+    }
+
+    private void takeActionWasAlreadyReported(@NotNull GuildMessageReceivedEvent event) {
+        // The user recently send the same scam already, and that was already reported and handled
+        addScamToHistory(event);
+
+        boolean shouldDeleteMessage = MODES_WITH_IMMEDIATE_DELETION.contains(mode);
+        if (shouldDeleteMessage) {
+            deleteMessage(event);
+        }
+    }
+
+    private void takeAction(@NotNull GuildMessageReceivedEvent event) {
+        switch (mode) {
+            case OFF -> throw new AssertionError(
+                    "The OFF-mode should be detected earlier already to prevent expensive computation");
+            case ONLY_LOG -> takeActionLogOnly(event);
+            case APPROVE_FIRST -> takeActionApproveFirst(event);
+            case AUTO_DELETE_BUT_APPROVE_QUARANTINE -> takeActionAutoDeleteButApproveQuarantine(
+                    event);
+            case AUTO_DELETE_AND_QUARANTINE -> takeActionAutoDeleteAndQuarantine(event);
+            default -> throw new IllegalArgumentException("Mode not supported: " + mode);
+        }
+    }
+
+    private void takeActionLogOnly(@NotNull GuildMessageReceivedEvent event) {
+        addScamToHistory(event);
+        logScamMessage(event);
+    }
+
+    private void takeActionApproveFirst(@NotNull GuildMessageReceivedEvent event) {
+        addScamToHistory(event);
+        logScamMessage(event);
+        reportScamMessage(event, "Is this scam?", createConfirmDialog(event));
```

Comment:
```suggestion
        reportScamMessage(event, "Is this scam?", List.of(createConfirmDialog(event)));
```
And the calling methods create a List of it.

Reply:
That defeats the purpose of the helper method. Its supposed to return exactly whats needed for the reportScamMessage call, to simplify the call-site. So I would rather rename it to `createActionRowsForConfirmDialog` to prevent what you are suggesting.

Reply:
🤷

Reply:
That said, why should the `reportScamMessage` take a list? It could as well just either take a single confirm dialog or nothing. It is domain-specific, doesnt need to be abstract.

Reply:
Don't like that either since it creates an action row, not action rows.
But w/e, you can undo the changes if you want.